### PR TITLE
WINC-551: Cluster autoscaling with Windows machines

### PR DESCRIPTION
**Pull request by saifshaikh48**
This PR adds a README section documenting how to enable
autoscaling of Windows MachineSets managed by WMCO.

### CODE REVIEW

**Comment by aravindhp** [comment]
/cherry-pick community-4.7

Just to check if the e2e tests are passing.

File: README.md [hunk 1]
```diff
@@ -146,6 +146,34 @@ WMCO is not responsible for Windows operating system updates. The cluster admini
...
+## Applying autoscaling to WMCO-managed Windows Machines
+
+There are two components to managing resources using autoscaling: a ClusterAutoscaler and MachineAutoscalers.
+The cluster autoscaler is capable of automatically adjust the size of the cluster based on node resource usage
+and overall deployment needs, within specific cluster constraints. Machine autoscalers modify machine sets
+to scale up and down the number of machines.
+Multiple MachineAutoscalers can be deployed to manage multiple MachineSets.
+Each machine autoscaler object should be mapped to a single machine set.
+
+- Define and deploy a [ClusterAutoscaler](https://docs.openshift.com/container-platform/4.6/machine_management/applying-autoscaling.html#configuring-clusterautoscaler).
```
Comment (jrvaldes) [inline README.md:158]: for my own education; does it only applies to version `4.6`  ? 

Otherwise, `latest` may help. See:  https://docs.openshift.com/container-platform/latest/machine_management/applying-autoscaling.html#configuring-clusterautoscaler

Comment (saifshaikh48) [reply README.md:158]: Nope, you are correct -- also applies to 4.7/4.8. I'll update to latest.

File: README.md [hunk 2]
```diff
@@ -146,6 +146,34 @@ WMCO is not responsible for Windows operating system updates. The cluster admini
...
+- Ensure a MachineSet resource is deployed for Windows machines (see spec in [Usage section](https://github.com/openshift/windows-machine-config-operator#usage)).
+- Create a MachineAutoscaler resource, referencing a MachineSet resource. Replace *\<windows-machineset-name\>*
```
Comment (mansikulkarni96) [inline README.md:160]: <windows-machineset-name\> does not appear to be in code format in the rich-diff.

File: README.md [hunk 3]
```diff
@@ -146,6 +146,34 @@ WMCO is not responsible for Windows operating system updates. The cluster admini
...
+The cluster autoscaler is capable of automatically adjusting the size of the cluster based on node resource usage
+and overall deployment needs, within specified constraints. Machine autoscalers modify machine sets
+to scale up and down the number of machines.
+Multiple MachineAutoscalers can be deployed to manage multiple MachineSets.
+Each machine autoscaler object should be mapped to a single machine set.
+
+- Define and deploy a [ClusterAutoscaler](https://docs.openshift.com/container-platform/latest/machine_management/applying-autoscaling.html#configuring-clusterautoscaler).
+- Ensure a MachineSet resource is deployed for Windows machines (see spec in [Usage section](https://github.com/openshift/windows-machine-config-operator#usage)).
+- Create a MachineAutoscaler resource, referencing a MachineSet resource. Replace *\<windows-machineset-name\>*
+  in the template below with the name of Windows MachineSet that was created, then deploy.
```
Comment (mansikulkarni96) [inline README.md:161]: maybe rephrase to specifics -> *name of Windows MachineSet that was created in the previous step, then deploy the MachineAutoscaler resource.

**Review by mansikulkarni96 — COMMENTED** [review]
Thanks for working on this @saifshaikh48 
LGTM, just some nits to consider.

File: README.md [hunk 4]
```diff
@@ -146,6 +146,34 @@ WMCO is not responsible for Windows operating system updates. The cluster admini
 creating the VMs and hence, the cluster administrator is responsible for providing an updated image. The cluster 
 administrator can provide an updated image by changing the image in the MachineSet spec.
 
+## Applying autoscaling to WMCO-managed Windows Machines
```
Comment (sebsoto) [inline README.md:149]: How does this process differ from Linux Machines?

If it doesn't it might make more sense to just point to the proper docs so we have less documentation to maintain

File: README.md (hunk 2 above)
Comment (saifshaikh48) [reply README.md:160]: Alright, I will change from _italicized_ to `code format`

File: README.md (hunk 4 above)
Comment (saifshaikh48) [reply README.md:149]: Great point... there is no difference between them. In that case would we want a short section that just points to the official docs, or delete this PR altogether?

Comment (sebsoto) [reply README.md:149]: I think that comes down to the reason this PR is being created. Is this something that we feel we need to convey to customers? Basically, is this a common question that we are getting? If so lets have a short section pointing to the official docs. If not? The section might not be needed. But what about the case where we have Windows only customers who are using only using this readme/the csv for MachineSet configuration? A short section makes sense. All that considered I think having a short section is better choice.

The other thing we have to think about is whether or not we are properly covering this in our tests. What is the underlying mechanism that autoscaling needs? Is it `oc adm top`? If so maybe the better way to approach this is adding an e2e test for interacting with a Windows node with `oc adm top`. If we arent testing this can we really promote this functionality?

**Comment by aravindhp** [comment]
>If so we might need to add an e2e test for interacting with a Windows node with oc adm top. If we arent testing this can we really promote this functionality?

While I am fine adding a test to see if `oc adm top` works, it is not tenable to for us to test very k8s feature out there in our e2e tests.

File: README.md (hunk 4 above)
Comment (saifshaikh48) [reply README.md]: Moved forward with a short section calling out our support + OpenShift docs.

Also, clarified the underlying mechanism [on the story](https://issues.redhat.com/browse/WINC-551?focusedCommentId=16205716&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16205716).

File: test/e2e/metrics_test.go [hunk 5]
```diff
@@ -287,3 +289,25 @@ func (tc *testContext) ensureMonitoringIsEnabled() error {
...
+	for _, winNode := range gc.nodes {
+		t.Run(winNode.Name, func(t *testing.T) {
+			cmd := exec.Command("oc", "adm", "top", "node", winNode.Name)
+			out, err := cmd.Output()
+			require.NoErrorf(t, err, "failed to get result of `%s`: %s", cmd, out)
+
+			// parse output for unreadable values. When metrics are not configured properly, values
+			// will be <unknown> as per https://bugzilla.redhat.com/show_bug.cgi?id=1920903
+			require.NotContains(t, string(out), "<unknown>")
+			assert.NoError(t, err, "could not retreive resource usage statistics for node %s", winNode.Name)
```
Comment (sebsoto) [inline test/e2e/metrics_test.go:310]: Looks like you're checking the same error twice

Instead try:
```
require.NoErrorf(t, err, "whichever error msg you think fits better")
assert.NotContains(t, string(out), "<unknown>")

```

**Review by sebsoto — CHANGES_REQUESTED** [review]
Thanks for making these changes @saifshaikh48

File: test/e2e/metrics_test.go (hunk 5 above)
Comment (sebsoto) [reply test/e2e/metrics_test.go:310]: Duplicate check was:
`require.NoErrorf(t, err, "failed to get result of `%s`: %s", cmd, out)`

**Comment by aravindhp** [comment]
/retitle WINC-551: Cluster autoscaling with Windows machines

File: test/e2e/metrics_test.go [hunk 6]
```diff
@@ -30,6 +31,7 @@ func testMetrics(t *testing.T) {
 	t.Run("Windows Exporter configuration validation", testWindowsExporter)
 	t.Run("Prometheus configuration validation", testPrometheus)
 	t.Run("Windows Prometheus rules validation", testWindowsPrometheusRules)
+	t.Run("Windows Node resource usage statistics validation", testNodeResourceUsage)
```
Comment (aravindhp) [inline test/e2e/metrics_test.go:34]: I suggest making this `Windows Node resource usage info validation` given how you have described the test

File: test/e2e/metrics_test.go [hunk 7]
```diff
@@ -287,3 +289,24 @@ func (tc *testContext) ensureMonitoringIsEnabled() error {
...
+func testNodeResourceUsage(t *testing.T) {
+	_, err := NewTestContext()
+	require.NoError(t, err)
+
+	// Need at least one Windows node to run these tests, throwing error if this condition is not met
+	require.Greater(t, len(gc.nodes), 0, "test requires at least one Windows node to run")
+
+	for _, winNode := range gc.nodes {
+		t.Run(winNode.Name, func(t *testing.T) {
+			cmd := exec.Command("oc", "adm", "top", "node", winNode.Name)
```
Comment (aravindhp) [inline test/e2e/metrics_test.go:303]: Should we make this an API call?

File: test/e2e/metrics_test.go [hunk 8]
```diff
@@ -7,6 +7,7 @@ import (
 	"fmt"
```
Comment (aravindhp) [inline test/e2e/metrics_test.go:7]: Given the test has nothing to do with cluster autoscaling, this should be a separate PR

File: test/e2e/metrics_test.go [hunk 9]
```diff
@@ -287,3 +289,24 @@ func (tc *testContext) ensureMonitoringIsEnabled() error {
...
+
+	for _, winNode := range gc.nodes {
+		t.Run(winNode.Name, func(t *testing.T) {
+			cmd := exec.Command("oc", "adm", "top", "node", winNode.Name)
+			out, err := cmd.Output()
+			require.NoError(t, err, "could not retreive resource usage statistics for node %s", winNode.Name)
+
+			// parse output for unreadable values. When metrics are not configured properly, values
+			// will be <unknown> as per https://bugzilla.redhat.com/show_bug.cgi?id=1920903
+			assert.NotContains(t, string(out), "<unknown>")
```
Comment (aravindhp) [inline test/e2e/metrics_test.go:309]: Is this the only error output that is possible?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this @saifshaikh48. Please address my comments.

File: README.md [hunk 10]
```diff
@@ -131,6 +131,14 @@ Usage:
 ./hack/machineset.sh apply/delete    # to create/delete MachineSet directly on cluster
 ```
 
+## Applying autoscaling to WMCO-managed Windows instances
+
+Cluster autoscaling is supported for Windows instances. 
+
+- Define and deploy a [ClusterAutoscaler](https://docs.openshift.com/container-platform/latest/machine_management/applying-autoscaling.html#configuring-clusterautoscaler).
+- Ensure a MachineSet resource is deployed for Windows machines (see spec in [Usage section](https://github.com/openshift/windows-machine-config-operator#usage)).
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply README.md:139]: I wanted to emphasize that autoscaling requires a MachineSet be present. I will change this to `Create a Windows node through a MachineSet` and that way it will stay relevant when BYOH support is introduced.

[the remark this answers is not in the document]
Comment (aravindhp) [reply README.md:139]: I am fine with that.

File: test/e2e/metrics_test.go (hunk 7 above)
Comment (saifshaikh48) [reply test/e2e/metrics_test.go:303]: This may be the way forward if we want to avoid 
> test very k8s feature out there in our e2e tests

Updated this test case in [a new PR](https://github.com/openshift/windows-machine-config-operator/pull/453). Please let me know if it is better.

Comment (aravindhp) [reply test/e2e/metrics_test.go:303]: Please remove this commit from this PR so that we can get it merged.

File: test/e2e/metrics_test.go (hunk 9 above)
Comment (saifshaikh48) [reply test/e2e/metrics_test.go:309]: This is the only error output I (and QE) has seen. I think 0% usage would also be worrisome as _some_ resources are always used by kubelet etc.

File: test/e2e/metrics_test.go (hunk 7 above)
Comment (saifshaikh48) [reply test/e2e/metrics_test.go:303]: On second thought, the test may be robust if we also included this logic/asserts as `oc adm top node` checks current resource usage stats and the API call checks allocatable resource stats. Both are useful to be checked in my opinion. Thoughts?

Comment (aravindhp) [reply test/e2e/metrics_test.go:303]: Like I said before, please remove this commit from this PR and move this discussion to #453

**Comment by mansikulkarni96** [comment]
Thanks for working on this @saifshaikh48 
/lgtm

**Review by aravindhp — APPROVED** [review]
Thanks for working on this
/approve

File: README.md [hunk 11]
```diff
@@ -146,6 +146,15 @@ WMCO is not responsible for Windows operating system updates. The cluster admini
 creating the VMs and hence, the cluster administrator is responsible for providing an updated image. The cluster 
 administrator can provide an updated image by changing the image in the MachineSet spec.
 
+## Enabled features
+
+### Autoscaling Windows nodes
```
Comment (aravindhp) [inline README.md:151]: @EricPonvelle is there a way to call out that this feature will be enabled for Windows nodes in our 2.0.1 release? Do you release note it? Or add it to the [Cluster Autoscaling section](https://docs.openshift.com/container-platform/4.7/machine_management/applying-autoscaling.html)? Anything works for us as long as it is called out.

/cc @Anandnatraj

File: README.md [hunk 12]
```diff
@@ -146,6 +146,15 @@ WMCO is not responsible for Windows operating system updates. The cluster admini
...
+Cluster autoscaling is supported for Windows instances. 
+
+- Define and deploy a [ClusterAutoscaler](https://docs.openshift.com/container-platform/latest/machine_management/applying-autoscaling.html#configuring-clusterautoscaler).
+- Create a Windows node through a MachineSet (see spec in [Usage section](https://github.com/openshift/windows-machine-config-operator#usage)).
+- Define and deploy a [MachineAutoscaler](https://docs.openshift.com/container-platform/latest/machine_management/applying-autoscaling.html#configuring-machineautoscaler), referencing a Windows MachineSet.
```
Comment (EricPonvelle) [inline README.md:156]: With this items, should I include these sections directly in the OpenShift docs version of this PR? Just want to make sure I have a full grasp of how to use this material correctly.

**Review by EricPonvelle — COMMENTED** [review]
lgtm - I added a few comments, more for my clarification.

I was also curious about what versions of OCP that these changes apply. Thanks for this!

File: README.md (hunk 12 above)
Comment (aravindhp) [reply README.md:156]: I don't think so given that there is nothing different for Windows here when compared to Linux. These steps are more for anyone trying to use the operator directly from the repo.

@saifshaikh48 please correct me if I am wrong.
